# eLabFTW - API workshop - Part 2 - Using curl

This notebook will guide you towards using the `curl` command line tool to make requests. But first, we need a key!

## **Chapter 1: Create an API Key for authentication**

As an eLabFTW user, you already have an account and can navigate to your settings to generate an API key for authentication.

For this part, we will need to login via the web interface to manage the api keys.

Here is the link to your local instance:

⇒ [API Workshop Instance](https://api-workshop.elab.one) ⇐

#### 1. **Log in and go to your `settings`**

![login settings](https://github.com/elabftw/api-workshop/blob/master/assets/1-login-settings.png?raw=true)

*(Click on the settings icon from the dropdown menu.)*

#### 2. **Navigate to the `API KEYS` tab**

![api keys navigation tab](https://github.com/elabftw/api-workshop/blob/master/assets/2-api-keys-tab.png?raw=true)


*(Find the API Key section where you can manage your keys.)*

#### 3. **Create a read/write key**
- Enter a name for the key (*e.g.* `api-for-test`).  
- Select the appropriate permission level. In this case, **read/write** would be appropriate for the course.
- Click **Generate API Key**.


![api key full tab](https://github.com/elabftw/api-workshop/blob/master/assets/3-api-key-full-tab.png?raw=true)

Congratulations! The API Key has been generated. **Make sure to save it somewhere safe** as you will use it often. We will need it now to send requests to the API.

![api key generated](https://github.com/elabftw/api-workshop/blob/master/assets/4-api-key-generated.png?raw=true)

## **Chapter 2: Make your first curl request with an API key**

### 1. **Setup your environment**

Before interacting with the API, the first step is to configure the necessary environment variables. These variables enable seamless communication with the API without requiring repeated manual input.

You will need to define the following:
- `API_URL`: The base URL of your eLabFTW instance.
- `API_KEY`:  The API key you recently generated.

In **Jupyter Lab**, you can use the `%env` magic command to set these environment variables, making them reusable by assigning them a key (name) instead of manually entering the full values each time.

#### Execute the commands

Use your **generated API Key** and your **API URL** to adapt the code below. For example:

%env API_URL=**yourdomain.elab/api/v2/**

%env API_KEY=**ADD YOUR KEY HERE**

Then, as explained in the [first section](./0-api-workshop-intro.ipynb#How-to-navigate-through-this-Notebook), **select the cell below** and click the play button!

In [ ]:
%env API_URL=https://api-workshop.elab.one/api/v2/
%env API_KEY=<ADD YOUR KEY HERE>

Now, verify the configuration by executing the following cell.

On success, this will display all environment variables containing **"API_"**, ensuring they are correctly stored and accessible for further API interactions.

In [ ]:
!env | grep API

<div class="alert alert-block alert-success"><b>You should see:</b>
    <p>API_URL=https://api-workshop.elab.one/api/v2/</p>
    <p>API_KEY=9-349864..............................................649</p></div>



### 2. **Now interact with the API!**

We will now use `curl` to have our first interaction with the API.
<b>curl</b> is a command-line tool used to make HTTP requests.
It will be completed by an `Authorization` header : when making API requests, you often need to send an authorization token in the headers to <b>authenticate</b>. We will provide it with the `-H` flag, like this:

```sh
curl $API_URL -H "Authorization: $API_KEY"" 
```

- `-H "Authorization: $API_KEY"`: This is the way of adding an HTTP header for authentication, using your API key.

<div class="alert alert-block alert-warning"><b>💡 Important:</b> Always use double quotes (") around <b>"Authorization: ..."</b> to ensure the terminal correctly interprets the $API_KEY variable. Using single quotes (') would pass the string literally instead of expanding the variable.</div>


#### **Send your first request!**

We will now complete the previous command with an [API Endpoint](./0-api-workshop-intro.ipynb#what-are-API-Endpoints?).

Let us run the cell below to send a `curl` request using the `GET` method to the `/info` endpoint. This will retrieve basic information about your eLabFTW instance.

[See here](https://doc.elabftw.net/api/v2/#/Info/get-info) for the documentation related to this endpoint.

In [ ]:
!curl "$API_URL"/info -H "Authorization: $API_KEY"

You should get this answer back from the server:

(if not, please refer to the [troubleshooting section](#Troubleshooting))

*Formatted JSON:*
~~~json
{
   "elabftw_version":"5.2.0-alpha",
   "elabftw_version_int":50200,
   "ts_balance":0,
   "ts_limit":0,
   "uploads_filesize_sum":116507,
   "uploads_filesize_sum_formatted":"113.78 KiB",
   "all_users_count":2,
   "active_users_count":2,
   "items_count":0,
   "teams_count":2,
   "experiments_count":1,
   "experiments_timestamped_count":0,
   "entities_timestamped_count_last_30_days":0
}
~~~

**Congratulations! You just did your first request to the eLabFTW API using curl.**

You can find a full list of the eLabFTW API's endpoints (*e.g.* `apikeys`, `experiments`, `config` *etc.*), with their different verbs (*e.g.* `GET`, `PATCH`) on this page:

https://doc.elabftw.net/api/v2/

![api swagger display full](https://github.com/elabftw/api-workshop/blob/master/assets/6-api-swagger-display.png?raw=true)

## **Chapter 3: Perform a full operation, from create to delete, with eLabFTW's API**

Now that you have all the tools at your disposal, let us perform some actions from A to Z with the API. We are going to **C**reate, **R**ead, **U**pdate and **D**elete an experiment (**CRUD**).

### 1. **Create an experiment with the POST verb**

![api experiments endpoint](https://github.com/elabftw/api-workshop/blob/master/assets/7-api-experiments-endpoint.png?raw=true)

We will start by creating an experiment with the command below (click to execute):

<details><summary>The command is in one line. <b>Click here</b> to see what the extended code looks like</summary>
    
~~~bash
curl -X 'POST' \
  'https://elab.local:3148/api/v2/experiments' \
  -H 'accept: */*' \
  -H "Authorization: $K" \
  -H 'Content-Type: application/json' \
  -d '{
  "body": "<h1>Section title</h1><p>Main text of resource</p>",
  "canread": "{\"base\": 20, \"teams\": [], \"users\": [1], \"teamgroups\": [1]}",
  "canwrite": "{\"base\": 20, \"teams\": [], \"users\": [1], \"teamgroups\": [1]}",
  "category": 3,
  "content_type": 1,
  "metadata": {
    "extra_fields": {
      "Information": {
        "type": "text",
        "value": "Lorem ipsum.",
        "required": true,
        "description": "This is a field of type: text"
      }
    }
  },
  "rating": 0,
  "status": 1,
  "tags": [
    "test",
    "confidential",
    "replication"
  ],
  "title": "Test replicating the results of Naomi",
  "template": -1
}'
~~~
</details>


In [15]:
# we add "-w" option to display the response status code.
!curl  -w "%{http_code}" -X POST "$API_URL"/experiments -H "accept: */*" -H "Authorization: $API_KEY" -H "Content-Type: application/json" -d '{"body":"<h1>Section title</h1><p>Main text of resource</p>","canread":"{\"base\": 20, \"teams\": [], \"users\": [1], \"teamgroups\": [1]}","canwrite":"{\"base\": 20, \"teams\": [], \"users\": [1], \"teamgroups\": [1]}","category":3,"content_type":1,"metadata":{"extra_fields":{"Information":{"type":"text","value":"Lorem ipsum.","required":true,"description":"This is a field of type: text"}}},"rating":0,"status":1,"tags":["test","confidential","replication"],"title":"Test replicating the results of Naomi - using curl","template":-1}'
# By default, there would be no output.

201

<div class="alert alert-block alert-success"><b>Success:</b> 201 Created: The resource was successfully created.</div>

### 2. **Read an experiment with the GET verb**

In this section, we use the tool `jq`. Install it with your package manager, or for **Google Colab** run the following cell.

In [ ]:
# only for google colab
!apt install jq

Let us read the content of an experiment. The command will retrieve a list of experiments from the eLabFTW API and extracts only the `id` and `title` fields from the response.

- The `curl` command sends a `GET` request to the API at `https://api-workshop.elab.one/api/v2/experiments`.
- The `-s` option makes the request silent, meaning it won’t show progress or error messages unless necessary.
- The `-H 'accept: application/json'` header ensures that the server returns the response in JSON format
- and the `-H "Authorization: $K"` header provides the required authentication using the API key stored in the `$K` environment variable.

The output of the `curl` command is piped (`|`) to `jq`, which is a tool for processing JSON data. The filter `.[] | {id, title}` tells `jq` to process each item in the returned JSON list and extract only the `id` and `title` fields from each experiment. The result is a list of objects containing just these two fields, making it easier to read and work with the relevant information without extra details from the full API response.

Click the cell below and execute to see the result :)

In [ ]:
!curl -s -X GET "$API_URL"/experiments -H 'accept: application/json' -H "Authorization: $API_KEY" | jq '.[] | {id, title}'

This should be your response from the server:

~~~json
{
  "id": 1,
  "title": "Test replicating the results of Naomi - using curl"
}
~~~

If you need not to filter the result for the id and title, you can of course remove the part with `| jq '....'`

We can now use the latest id, which is the experiment we created, to try and get **one specific experiment** and not all of them.

### 3. **Read one specific experiment with the GET verb**

The command below will retrieve one experiment and its information.

In [ ]:
!curl -s -X GET "$API_URL"/experiments/1 -H 'accept: application/json' -H "Authorization: $API_KEY" | jq '.'

### 4. **Update an experiment with the PATCH verb**

This command patches the experiment with the ID you provide.

Here, replace the **"X"** in "/experiments/<span class="alert-danger">X</span>" with an <b>id</b> from the previous <b>GET</b> request.


In [17]:
!curl -s -X PATCH "$API_URL"/experiments/X -H 'accept: application/json' -H "Authorization: $API_KEY" -H 'Content-Type: application/json' -d '{"title": "Updated the title through curl"}' | jq '.'

{
  "code": 400,
  "message": "Bad Request",
  "description": "Cannot check permissions without an id!"
}


<div class="alert alert-block alert-success"><b>Success:</b> 200 Success: The resource is returned with updated content.</div>



Should you see any other answer, please refer to the 👉 [troubleshooting section](#Troubleshooting).

### Delete an experiment with the DELETE verb

Deleting an experiment is a simple task. You need to provide the ID as the parameter, and the response will be empty on success, just like the POST verb.

In [ ]:
!curl -s -X 'DELETE' "$API_URL"/experiments/2 \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "Authorization: $API_KEY" | jq '.'

<div class="alert alert-block alert-success"><b>Success:</b> 204 No Content: The resource was successfully deleted, and no further content is returned.</div>

Congratulations on finishing the second part of this course ! If you want to improve even further your lab's workflow, you can proceed to the next part of this course and learn about **python** programming, to automate your actions.

➡️ See [last part](./2-api-workshop-python.ipynb) of this course.

# Troubleshooting

Here are some issues you could have encountered during this course.

### Error 400: Bad Request

~~~json
{
   "code":400,
   "message":"Bad Request",
   "description":"Invalid endpoint: available endpoints: apikeys, batch, config, idps, idps_sources, import, info, experiments, exports, items, experiments_templates, items_types, event, events, extra_fields_keys, favtags, team_tags, teams, todolist, unfinished_steps, users"
}
~~~

**Explanation:**
The error message indicates that the request was made to an invalid or non-existent endpoint. The API responded with a 400 Bad Request error, meaning the request is not correctly formatted or does not match an available resource.

The description provides a list of valid endpoints, such as `apikeys`, `config`, `experiments`, `users`, etc. This suggests that the issue is likely due to a typo in the endpoint name or an incorrect URL structure. To resolve this, on the next step we will use one of the listed endpoints in the request.

You can also receive a **400** error such as :
~~~json
{
  "code": 400,
  "message": "Bad Request",
  "description": "Cannot check permissions without an id!"
}
~~~

### Error 401: Authentication required

~~~json
{
    "code":401,
    "message":"Unauthorized",
    "description":"Authentication required"
}
~~~ 

**Explanation:**
This error occurs when the API key is incorrect or missing. The server rejects the request because authentication is required but not properly provided. Ensure that:

- Your API key is correctly set in the environment
- You are using double quotes around "Authorization: $API_KEY" in your curl request.

### Error 401: No corresponding API Key

~~~json
{
    "code":401,
    "description":"No corresponding API key found!"
}
~~~ 

**Explanation:**
This message indicates that the provided API key exists and is correctly formatted, but it does not match any valid key on this eLabFTW instance. Possible reasons include:

- The API key was created in a different eLabFTW instance.
- The key has been revoked or deleted.
- The key was not copied correctly when set in the environment.

🔹 Solution: Double-check that you're using an active API key from the correct eLabFTW instance.